### ISSUE TITLE
Complete the SemanticVersion work 

### ISSUE THREAD

**Issue by HemantMahawar** [PowerShell/PowerShell#2983]
- [ ] Support SemVer in #Requires (PowerShell version (-Version) as well as module version (-Modules))
- [ ] Allows cmdlet parameters to consume SemVer instead of System.Version (argument transformation maybe?)
  - [ ] Users can convert Semver to System.Version using that attribute before passing to the cmdlet parameter
  - [ ] Cmdlet author can mark the attribute on the parameter and users can pass Semver string

=======================
**User Experience:**
- On disk, module versions presents - 1.0.0, 1.0.0-alpha17, 1.0.1 
- Result:

  | Command  | Expected Output | Comments |
  | -------------| ------------------ | -------------|
  | Import-Module -Version 1.0.2 | Error |  Version is alias to min version |
  | Import-Module -MinimumVersion 1.0.0 | Module with version 1.0.1 should be loaded |   |
  | Import-Module -Version 1.0.0-alpha15 | Module with version 1.0.1 should be loaded |  |
  | Import-Module -Name | Module with version 1.0.1 should be loaded | Latest version of module should be loaded |
  | Import-Module -RequiredVersion 1.0.0 | Module with version 1.0.0 should be loaded |  |

Note that 1.0.0 is higher than 1.0.0-alpha17, so if 1.0.1 doesn't exist, then `Import-Module` would load 1.0.0.

**Comment by SteveL-MSFT** [comment]
We should wait for corefx support of semver (https://github.com/dotnet/corefx/issues/13526) rather than continue to have our implementation

**Comment by iSazonov** [comment]
There is Set-StrictMode. Is it out of the enhancement?
```powershell
 gcm -ParameterName Version

CommandType     Name                                               Version    Source
-----------     ----                                               -------    ------
Cmdlet          Import-Module                                      7.0.0.0    Microsoft.PowerShell.Core
Cmdlet          Set-StrictMode                                     7.0.0.0    Microsoft.PowerShell.Core
```

Do we agree with breaking change that 1.0.0.1 is not SemVer?

**Comment by SteveL-MSFT** [comment]
Moving out as I think there's a chance .NET 5 will have semver type

**Comment by tnieto88** [comment]
At what point does PowerShell stop waiting for .NET and use NuGet version to support both Semver and traditional versions in the same type?

**Comment by tnieto88** [comment]
How realistic is it that they will complete it in version 5 we've already been waiting 3 years for a resolution and it's been 3 months since the last comment. NuGet already has a working semver type.

**Comment by iSazonov** [comment]
> NuGet already has a working semver type

Core team is still not finished design. 

You could ask in https://github.com/dotnet/corefx/issues/13526.

**Comment by rkeithhill** [comment]
That issue is marked "up for grabs".  I wouldn't count on the CoreFX team implementing this.  Time to move on and come up with a solution of some sort for PowerShell IMO.

**Comment by vexx32** [comment]
I've looked at taking up that issue once in the past, and the roadblock I hit was about the same the other folks who've looked at it that I spoke to have also hit.

Essentially, it seems the bar has been set _far too high_ for a SemVer implementation in .NET Core / .NET 5. Everyone who sees that someone is asking for a concrete SemVer implementation seems to question it, and by far the most complex question is "given a random SemVer, how do I tell what version of SemVer they're using?"

And there isn't really a clear answer; it's not a problem SemVer is especially well-designed to be able to handle.

And thus far it seems like the .NET Core team is content to wait until _everyone_ is happy with a given proposed specification before moving forward, which... quite **literally**  by definition will never happen. Until .NET Core change their stance and at least allow a _first_ SemVer implementation in, there will never be one in .NET Core, and there will never be improvements made to one if they don't have a base to work with.

**Comment by iSazonov** [comment]
> That issue is marked "up for grabs". 

@rkeithhill It was removed.

**Comment by SteveL-MSFT** [comment]
We want to avoid a situation where we have competing semver types.

**Comment by tnieto88** [comment]
I understand that, but I would like time frame on how much longer PowerShell will wait on CoreFx before doing their own thing. One, two, five years?

**Comment by SteveL-MSFT** [comment]
If it appears that it won't be fixed in .NET 5, we'll probably have to fix it ourselves within our code base.

**Comment by ThomasNieto** [comment]
Looks like there is already a `[System.Management.Automation.SemanticVersion]` type used in `$PSVersionTable` can that type be used until .NET releases their semver type? It should at least be used in the module system so PowerShellGet isn't trying to force semver into a system that doesn't support it.

**Comment by vexx32** [comment]
Yeah it's pretty janky the way we currently handle it there. @SteveL-MSFT are there plans to implement semver in PSGet v3?

**Comment by ThomasNieto** [comment]
@SteveL-MSFT / @SydneyhSmith what are the plans for the module system and PowerShellGet supporting both legacy `[System.Version]` and .NET `[SemanticVersion]` if/when that ever gets released?

PowerShellGet v3 is facing multiple issues trying to support both version schemes:

https://github.com/PowerShell/PowerShellGet/issues/269 - multiple types are returned for the Version property, which cannot be sorted
https://github.com/PowerShell/PowerShellGet/issues/264 - `[Version]` is used stripping off `Prerelease` label
https://github.com/PowerShell/PowerShellGet/issues/261 - full semver folder used causing `Import-Module` to fail.
https://github.com/PowerShell/PowerShellGet/issues/78 - cannot upgrade from prerelease to stable with binary files in use

`PackageManagement`/`PowerShellGet` v2 used `[string]` for the `Version` property and it did not work well since there was no actual version type sorting (comparison) done as described in: https://github.com/OneGet/oneget/issues/432

My opinion is that a single type that supports both version schemes would be the best. `[NuGet.Versioning.NuGetVersion]` is a type that meets that requirement which was recommend by @Jaykul in https://github.com/PowerShell/PowerShell-RFC/pull/185#discussion_r294625087

Whatever is decided both PowerShell and PowerShellGet have to use the same solution or we'll end up with the current system of PowerShellGet trying to force semver into a module system that doesn't support it. This problem has to be solved in PowerShellGet by the time v3 goes stable.

**Comment by iSazonov** [comment]
PowerShell already has 
```public sealed class SemanticVersion : IComparable, IComparable<SemanticVersion>, IEquatable<SemanticVersion>```

**Comment by ThomasNieto** [comment]
@iSazonov during the August community call the team said they didn't want to use that class since when .NET released one there would duplicate competing types.

**Comment by iSazonov** [comment]
@ThomasNieto Yes, but it could be used as _an implementation detail_.
It is old history. We wait .Net very long. I hope they implement the class in 6.0. If no we _can_ use custom implementation for internal use.

**Comment by ThomasNieto** [comment]
The part I'm most interested about is how multiple version schemes will be supported on the `[PSModuleInfo]` type or would multiple types be needed. Right now, the `Version` property is a `[System.Version]` type. When semver is officially supported by the module system then how would `*-Module` cmdlets handle returning a type with `Version` property having different types? For example, `[System.Version]` for legacy and `[SemanticVersion]` in the SMA or .NET namespace.

**Comment by iSazonov** [comment]
> The part I'm most interested about is how multiple version schemes will be supported

It is a big headache in global and this explains why .Net team does not implement the simple class jet.

Obviously both types will have to coexist and two cases exist - (1) exposing (like in Get-Module output), (2) using/implementing (like how adopt scripts for new Get-Module output). It is true for C# too.
Obviously there is not an universal solution but as a helper solution these types could be enhanced (both or one) with new field to reference original (add SemantecVersion property in Version type or Version property in SemantecVersion type)

**Comment by SeeminglyScience** [comment]
Yeah... ideally we'd wait for .NET but it doesn't look like that's in the cards.

Here's what I'd do:

1. Copy the best implementation currently out there (license permitting)
1. Put it in it's own assembly entirely
1. Ship that single type assembly on nuget targeting netstandard2.0 or even net452 so it can be used in *anything* and everything
1. Maybe even ship a module with `ConvertTo-SemanticVersion` and possibly `ConvertFrom-SemanticVersion` so it's easy to bundle with script-only modules

That assembly would serve as a default implementation.  PowerShell can ship it's own version of that assembly, if desired, that just redirects with `TypeForwardedTo` to a SMA.dll implementation.

**Comment by SeeminglyScience** [comment]
And just to clarify, I am aware of `SMA.SemanticVersion`.  Problem is you can't use it everywhere.  Usually I'm fine with keeping something only in Core to help adoption, but I don't think we can really wait until we can all get away with not targeting Windows PowerShell for this one.  I've already started to see questions about how to deal with semantic versions bleed over into reddit (usually the last place in the PS community to see something adopted).  I think I've seen ~3 threads in the last two weeks alone?  Might not seem like it a lot, but it's unusual to see that many repeats that quickly there.

This should be figured out before PSGet 3 GA imo.

**Comment by Jaykul** [comment]
I guess this will never be done then?

**Comment by ThomasNieto** [comment]
> If it appears that it won't be fixed in .NET 5, we'll probably have to fix it ourselves within our code base.

@SteveL-MSFT its been almost 4 years since this comment. .NET 5, 6, 7 and 8 are not going to implement semantic versioning. How much longer does PowerShell need to wait, ideally this would go out with PSResourceGet so both the engine and PSResourceGet can use semantic versioning without the pitfalls of half supporting it.

**Comment by ThomasNieto** [comment]
@StevenBucher98 can we get the working group to reevaluate this? It's been open for years now without any indication from dotnet that they will include this type by default.

**Comment by JamesWTruher** [comment]
The WG discussed this and in the absence of underlying support from the .NET team, we don't believe we should undertake any effort in this area.

**Comment by MariusStorhaug** [comment]
Feel free to look at PSSemVer :)

**Comment by ThomasNieto** [comment]
@JamesWTruher if that is the case please remove the SemanticVersion support from module versions as it the half support we have today hurts the PowerShell experience. My opinion is that you fully support it or remove it from PowerShell and PSResourceGet.

**Comment by ThomasNieto** [comment]
Brought this up with the following points in the PS community call.

The background of this issue is to complete the Semantic Versioning support in the engine due to PSResourceGet

* The WG decision does not have any context, address any issues, or consider any alternate solutions.
	As such indicates Semantic Versioning is not supported in PowerShell and by extension PSResourceGet and should be removed

* The scope of work is allowing Semantic versioning in module cmdlets
	which could be accomplished by adding a (65KB) dependency on NuGet.Versioning 

* Here are some of the issues when PSResourceGet supports Semantic Versioning while the engine does not
	Issue 1# - Prerelease binary modules cannot be updated to stable.
		The largest impact of this is PSReadLine as users who install the prerelease version will have to under go workarounds to install the stable version; thus hurting testing adoption

	Issue 2# - Prerelease and stable versions cannot be installed at the same time

	Issue 3# - Formatter displays a fake Prerelase property causing user confusion

* PowerShell should either fully support Semantic versioning in the engine or remove it completely and in PSResourceGet

* My concern about WG verdicts and lack of transparancy could be addressed by
	WG meetings being recorded, having meeting minutes, and a publicly accessable vote with accompaning concurring or dissenting opinions.

	This would help users know their issue has been earnestly considered 
	
* I want the PowerShell committee to review the future of Semantic Versioning as PSResourceGet should not support a versioning scheme that PowerShell cannot

**Comment by ThomasNieto** [comment]
@SydneyhSmith please reopen this issue as was discussed on the community call.

**Comment by ThomasNieto** [comment]
@SydneyhSmith the issue closed again because the tags are still on it.

**Comment by SeeminglyScience** [comment]
> * The scope of work is allowing Semantic versioning in module cmdlets
>   which could be accomplished by adding a (65KB) dependency on NuGet.Versioning

This was discussed, but adding a dependency can cause more issues than just the size. `Newtonsoft.Json` for example is a dependency we took that is a large pain point for a lot of folks.

That doesn't mean we can't add dependencies, it just raises the bar for impact significantly.

> * Here are some of the issues when PSResourceGet supports Semantic Versioning while the engine does not
>   Issue 1# - Prerelease binary modules cannot be updated to stable.
>   The largest impact of this is PSReadLine as users who install the prerelease version will have to under go workarounds to install the stable version; thus hurting testing adoption
>   Issue 2# - Prerelease and stable versions cannot be installed at the same time

PSReadLine now increments the `Revision` (or maybe `Patch`?) number for the version when it releases a new preview or GA. It's not perfect, but it is a workaround that works well.

That said, keep in mind this issue is *very* broad. The WG is only saying that we cannot commit to fixing every issue with the semver type *and* also everything around it - at least without support from the BCL. For example, I do personally think it would be smart to at least change module import look up to look for semver folders, even if PSRG won't be able to save them to that format for likely a few years. Still should get the ball rolling on it, I've opened #21371 to track that.

>   Issue 3# - Formatter displays a fake Prerelease property causing user confusion

FWIW I very much dislike fake properties in formatting in general, and I agree that at the very least an ETS property should be added.  Feel free to open an issue for that if you'd like

> * My concern about WG verdicts and lack of transparency could be addressed by
>   WG meetings being recorded, having meeting minutes, and a publicly accessible vote with accompanying concurring or dissenting opinions.
>   This would help users know their issue has been earnestly considered

I don't disagree, but it is also a bit of a give and take there. The more public it is, the harder it is to have a candid discussion - which means it takes longer and we get to even less issues than we do now.

It's the same reason why we often can't explain in great detail everything that was discussed, because writing all of that up takes a lot of time.

It's definitely something we're thinking about how to improve.

**Comment by hilari0n** [comment]
> > * The scope of work is allowing Semantic versioning in module cmdlets
> >   which could be accomplished by adding a (65KB) dependency on NuGet.Versioning
> 
> This was discussed, but adding a dependency can cause more issues than just the size. `Newtonsoft.Json` for example is a dependency we took that is a large pain point for a lot of folks.
> 
> That doesn't mean we can't add dependencies, it just raises the bar for impact significantly.

Uhm... So one of the reasons of not using `NuGet.Versioning` is that it would be a new/added reference.
From what I see, `NuGet.Versioning` is already delivered with `Microsoft.PowerShell.PSResourceGet` (at least since version 1.0.0), which is delivered with PowerShell.
And - as far as I know - it's quite improbable for it to go away, as it's one of the core mechanics of NuGet packages versioning, which are something, that PowerShell relies on (and should probably start embracing those `Nuget.*` packages, instead of building its own mechanics for NuGet handling).

**Comment by SeeminglyScience** [comment]
> Uhm... So one of the reasons of not using `NuGet.Versioning` is that it would be a new/added reference. From what I see, `NuGet.Versioning` is already delivered with `Microsoft.PowerShell.PSResourceGet` (at least since version 1.0.0), which is delivered with PowerShell.

True! But it does so in its own assembly load context which won't conflict. For PowerShell to fully embrace the type it would need to be registered and loaded at start up and in the default ALC. Even if we were able to lazily load it somehow (which we likely wouldn't be able to do due to it needing to be referenceable from script) we would still need to lock down in our assembly resolver which assembly is loaded.

**Comment by hilari0n** [comment]
> > And PowerShell is apparently based on .NET SDK (i.e., PowerShell release notes always reference a .NET SDK version, not e.g., .NET runtime version, quote from release notes 7.4.1 "Bump .NET 8 to version 8.0.101. Update .NET SDK and dependencies for v7.4.1"), which does contain `NuGet.Versioning`.
> 
> Bumping the SDK is required to get the latest runtime. `NuGet.Versioning` is not shipped with every dotnet project, it's just a compile time dependency

SDK-s are not required to get the latest runtime. Runtime is also distributed separately and you can use it that way.
Also, what latest runtime you would get from SDK 8.0.203, over the one from SDK 8.0.202? You can even use SDK 9.0 to build an application for the 8.0 runtime and it will work with that runtime (just as I'm using SDK 8.0 to build .NET apps for runtimes 6.0 and 7.0 along with ones for 8.0).
SDK is a compile-time dependency in a very similar way, as `NuGet.Versioning` is.

It's baffling, that I'd get the exact SDK version number used for building and packaging in the release notes, but the actual runtime version, which is packaged with PowerShell, is something I have to look up. It's also baffling, that PowerShell needs to use the self-contained model and deliver the runtime, so I have to have another copy of it on my machine, (and wait for separate security updates for PowerShell, when there are runtime security updates available).

Going back to having the `NuGet.Versioning` NuGet in PowerShell:
It's a much less breaking change to have it there and even have it fixed to some version (and have potential issues with modules wanting to rely on a different one), than changes like e.g., dropping the `NJsonSchema` NuGet in 7.4.0 completely (and breaking a lot of uses of `Test-Json`). If someone's OK with messing up how people work win Json in PowerShell, when switching from 7.3 to 7.4, but is not OK with adding `NuGet.Versioning` somewhere between Windows PowerShell 5.1 and PowerShell 7.4, then I can't grasp the thought process behind it.

**Comment by SeeminglyScience** [comment]
> SDK-s are not required to get the latest runtime. Runtime is also distributed separately and you can use it that way.

The SDK is what we use to build PowerShell. If you build a self contained project with a lower version of the SDK then you get a lower version of the runtime.

> Also, what latest runtime you would get from SDK 8.0.203, over the one from SDK 8.0.202?

Sorry, I don't know the specifics of every SDK version. I imagine that some of them are only tooling changes, but for example from 8.0.201 to 8.0.203 the runtime version is higher when compiling a self contained project.

> You can even use SDK 9.0 to build an application for the 8.0 runtime and it will work with that runtime (just as I'm using SDK 8.0 to build .NET apps for runtimes 6.0 and 7.0 along with ones for 8.0). 

Yeah for sure, you just can't use an older version of the SDK to build for a newer version.

> SDK is a compile-time dependency in a very similar way, as `NuGet.Versioning` is.

Yep! Exactly

> Going back to having the `NuGet.Versioning` NuGet in PowerShell: It's a much less breaking change to have it there and even have it fixed to some version (and have potential issues with modules wanting to rely on a different one), than changes like e.g., dropping the `NJsonSchema` NuGet in 7.4.0 completely (and breaking a lot of uses of `Test-Json`). If someone's OK with messing up how people work win Json in PowerShell, when switching from 7.3 to 7.4, but is not OK with adding `NuGet.Versioning` somewhere between Windows PowerShell 5.1 and PowerShell 7.4, then I can't grasp the thought process behind it.

I wasn't involved with that change and don't know much about it, but even if breaks were intentional there - that doesn't really impact the decision here. One break doesn't mean all breaks are warranted. Especially if a previous change had a bad outcome, if anything that would be an example of why not to do it, not why to do it more.

**Comment by daxian-dbw** [comment]
> It's baffling, that I'd get the exact SDK version number used for building and packaging in the release notes, but the actual runtime version, which is packaged with PowerShell, is something I have to look up.

This can be improved to link to the .NET release page of that version of .NET SDK. It will make it easy for you to find out what version of .NET runtime coming with it.

> It's also baffling, that PowerShell needs to use the self-contained model and deliver the runtime, so I have to have another copy of it on my machine

Then maybe you want to use the [PowerShell global tool](https://devblogs.microsoft.com/powershell/introducing-powershell-as-net-global-tool/) instead, which depends on the corresponding runtime to be already installed separately on your machine.

> dropping the NJsonSchema NuGet in 7.4.0 completely (and breaking a lot of uses of Test-Json)

Although that breaking change is not relevant to this discussion, I want to add some context about it:
1. This comment https://github.com/PowerShell/PowerShell/issues/18009#issuecomment-1235852077 was why we decided to accept that change.
1. We considered it a breaking change at the time in the sense that we were switching to a different library, but we didn't expect the behavior of `Test-Json` to be changed.

**Comment by SteveL-MSFT** [comment]
Keeping this open as the .NET issue is still open.  If they decide to never do it, we can revisit, but it doesn't make sense for PS7 to have it's own semver implementation with the risk of .NET having a different one and the two may not be compatible.

**Comment by SteveL-MSFT** [comment]
The main problem that might not be completely apparent is that if PS7 has a implementation of semver, then we need to support it indefinitely as the core platform.  Individual modules may do what they need for their cases, but other apps not specific to that domain would not take a dependency on it.  Although not ideal, we can continue to not have a semver type.

**Comment by ThomasNieto** [comment]
At this point I just want the module issues relating to semver prerelease modules fixed as that is the main issue.

From the discussion on this thread these are the non-starters:

- No competing semver types from PS and .NET
- No use of [S.M.A.SemanticVersion] as it violates competing types 
- No new library dependencies like [NuGet.Versioning] as version conflicts can/will happen and can't use ALC.
- No breaking changes such as changing property types or parameter types

Here is my proposal address fixing semver prerelease modules.

1. Complete #21371
2. Create a new parameter called VersionRange that accepts a `String`. This parameter would accept NuGet version range syntax like PSRG. This is needed as the existing version parameters are type [Version] and and would be a breaking change to change types. Another benefit is to align PSRG and PS into a common interface simplifying the number of parameters needed. The last and main benefit to this is that the parsing of the string can be handled internally not exposing a semver type to the user and handling both [System.Version] and semver.
3. Optional, move `MinimumVersion`, `MaximumVersion, `RequiredVersion` into a legacy parameter set (might not be able to with the amount of parameter sets today) or hide them for backwards compatibility
4. Add a `SemanticVersionString` with type of `String?` that is populated when the version can be parsed into a semver. This is to not expose a semver type to the user
5. Update PSRG to use semver prerelease folders when on PS Versions supporting semver prerelease. This should not be an issue as if the user requested the prerelease version on an older PS version than the current state is there if the new PS version is used then it would be in a different folder. Then if the user requested the stable version it would not run into the binary update issue and be placed into the correct version folder. This is not a breaking change as the user requests the version if they install the prerelease or stable versions.

Then once a .NET type is released.

1. Take dependency on semver .NET type
2. Update internal logic to use semver .NET type
3. Add `SemanticVersion` property using that semver .NET type

This proposal solves the main issues with the current state while not breaking any of the non-starters and leaving room in the future for a real .NET semver type.
